src/optimizer/optimizer.cpp

```diff
@@ -221,398 +220,106 @@ shared_ptr<GroupExpression> Optimizer::InsertQueryTree(
+      std::unique_ptr<parser::OrderDescription> &order_info,
+      std::vector<expression::AbstractExpression *> &output_exprs,
+      std::shared_ptr<PropertySet> &physical_props) {
+    // Extract output column
+    for (auto &expr : select_list) output_exprs.push_back(expr.get());
+
+    // Extract sort property
+    if (order_info != nullptr) {
+      std::vector<expression::AbstractExpression *> sort_exprs;
+      std::vector<bool> sort_ascending;
+      for (auto &expr : order_info->exprs) {
+        sort_exprs.push_back(expr.get());
+      }
+      for (auto &type : order_info->types) {
+        sort_ascending.push_back(type == parser::kOrderAsc);
+      }
+      if (!sort_exprs.empty())
+        physical_props->AddProperty(
+            std::make_shared<PropertySort>(sort_exprs, sort_ascending));
+    }
+  };
+
+  std::vector<expression::AbstractExpression *> output_exprs;
+  std::shared_ptr<PropertySet> physical_props = std::make_shared<PropertySet>();
+  switch (tree->GetType()) {
+    case StatementType::SELECT: {
+      auto select = reinterpret_cast<parser::SelectStatement *>(tree);
+      GetQueryInfoHelper(select->select_list, select->order, output_exprs,
+                         physical_props);
+      break;
+    }
+    case StatementType::INSERT: {
+      auto insert = reinterpret_cast<parser::InsertStatement *>(tree);
+      if (insert->select != nullptr)
+        GetQueryInfoHelper(insert->select->select_list, insert->select->order,
+                           output_exprs, physical_props);
+      break;
+    }
+    default:;
+  }
 
-unique_ptr<planner::AbstractPlan> Optimizer::OptimizerPlanToPlannerPlan(
-    shared_ptr<OperatorExpression> plan, PropertySet &requirements,
-    vector<PropertySet> &required_input_props,
-    vector<unique_ptr<planner::AbstractPlan>> &children_plans,
-    vector<ExprMap> &children_expr_map, ExprMap *output_expr_map) {
-  OperatorToPlanTransformer transformer;
-  return transformer.ConvertOpExpression(plan, &requirements,
-                                         &required_input_props, children_plans,
-                                         children_expr_map, output_expr_map);
+  return QueryInfo(output_exprs, physical_props);
 }
 
 unique_ptr<planner::AbstractPlan> Optimizer::ChooseBestPlan(
-    GroupID id, PropertySet requirements, ExprMap *output_expr_map) {
-  Group *group = memo_.GetGroupByID(id);
-  shared_ptr<GroupExpression> gexpr = group->GetBestExpression(requirements);
+    GroupID id, std::shared_ptr<PropertySet> required_props,
+    std::vector<expression::AbstractExpression *> required_cols) {
+  Group *group = metadata_.memo.GetGroupByID(id);
+  LOG_TRACE("Choosing with property : %s", required_props->ToString().c_str());
+  auto gexpr = group->GetBestExpression(required_props);
 
   LOG_TRACE("Choosing best plan for group %d with op %s", gexpr->GetGroupID(),
             gexpr->Op().name().c_str());
 
```

Comment:
Please break up this function here and make them two separate functions. The first choose the best expression given the required properties, and the second transform that into an execution plan. That would be semantically cleaner and easier for others to use.

Reply:
In general, I agree that we should separate the logic of different functionalities. But in this case, generating the operator expression first seems to introduce some more complexity. This is because generating the input columns for the plan needs the table aliases set in the `Group` to know which column is in which child group, thus we may want to use the `GroupExpression`. If we first generate an `OperatorExpression` tree, then generate the plan based on the tree, it could be a bit difficult to efficiently generate column information for the plan.

Reply:
Okay, that's fine. Let's leave this thing here then.